jupyterhub/app.py

```diff
@@ -737,6 +739,55 @@ def _domain_default(self):
             return ''
         return urlparse(self.subdomain_host).hostname
 
+    subdomain_hook = Union(
+        [Callable(), Unicode()],
+        default_value="legacy",
+        config=True,
+        help="""
+        Hook for constructing subdomains for users and services.
+
+        Default: 'legacy'
+        Use 'idna' to select a newer, more robust scheme,
+        which will become the default in the future.
+
+        A custom subdomain hook should have the signature:
+
+        def subdomain_hook(name, domain, kind) -> str:
+            ...
```

Comment:
I don't understand this config so well from the docstring alone, I think the following could be concrete suggestions:

- Mention that `"legacy"` and `"idna"` string options refer to _pre-defined hook functions_
- Clarify what the pre-defined hook function does in this docstring or link out to read about them elsewhere.
- I think maybe this config is relevant only if another config option is enabled, is it? I've never worked with a jupyterhub providing subdomains for each user, and figure that it must be opted-in one way or another.

Reply:
All great points that should be addressed in the docstring, thanks!